src/cargo/core/compiler/mod.rs

```diff
@@ -862,8 +862,14 @@ fn build_base_args(
     add_allow_features(cx, cmd);
 
     if !test {
-        for crate_type in crate_types.iter() {
-            cmd.arg("--crate-type").arg(crate_type.as_str());
+        if let Some(crate_types) = cx.bcx.rustc_crate_types_args_for(unit) {
```

Comment:
It looks like the `prefer_dynamic` below is perhaps looking at the wrong crate_types when this is set.